nca/Parsers/IstioGatewayPolicyGenerator.py

```diff
@@ -0,0 +1,298 @@
+        conn_cube["dst_ports"] = dest.ports
+        conns = ConnectionSet()
+        conns.add_connections(self.protocol_name, ConnectivityProperties.make_conn_props(conn_cube))
+        conn_cube.update({"src_peers": source_peers, "dst_peers": dest.pods, "protocols": self.protocols})
+        opt_props = ConnectivityProperties.make_conn_props(conn_cube)
+        if is_ingress:
+            return GatewayPolicyRule(source_peers, conns, opt_props)
+        else:
+            return GatewayPolicyRule(dest.pods, conns, opt_props)
+
+    @staticmethod
+    def create_deny_rule(source_peers, dst_peers):
+        """
+        Create a rule for representation of the denied connections from between given mesh and DNS nodes
+        :return: the resulting GatewayPolicyRule
+        """
+        opt_props = ConnectivityProperties.make_conn_props_from_dict({"src_peers": source_peers,
+                                                                      "dst_peers": dst_peers})
+        return GatewayPolicyRule(dst_peers, ConnectionSet(True), opt_props)
+
+    def create_gtw_to_mesh_and_deny_policies(self, vs, route, route_cnt, gtw_to_hosts, used_gateways):
+        """
+        Create internal policies representing connections from gateways to mesh,
+            described by the given http/tls/tcp route.
+        When relevant, create also deny policies from mesh to the mentioned DNS entries.
+        :param VirtualService vs: the virtual service holding the given route
+        :param Route route: the given http/tls/tcp route from which policies should be created
+        :param int route_cnt: the index (starting from 1) if the given route inside this virtual service routes
+        :param dict(Gateway, MinDFA) gtw_to_hosts: a map from gateways to hosts relevant for this route
+        :param set(Gateway) used_gateways: a set of used gateways, to be updated by adding gateways
+               referenced by these routes
+        :return: list[GatewayPolicy] the resulting list of policies
+        """
+        result = []
+        this_route_conn_cube = self.init_route_conn_cube(route)
+        for gtw, host_dfa in gtw_to_hosts.items():
+            # Modeling connections from ingress gateway nodes to internal service nodes (Ingress flow) or
+            # from egress gateway nodes to external service nodes (DNS nodes) (Egress flow).
+            if route.all_sni_hosts_dfa:
+                host_dfa &= route.all_sni_hosts_dfa
+                if not host_dfa:
+                    continue
+            this_route_conn_cube["hosts"] = host_dfa
+            used_gateways.add(gtw)
+            allow_policy = self.init_gtw_to_mesh_policy(vs, route_cnt, gtw)
+            deny_policy = None
+            if gtw.type == Gateway.GatewayType.Egress:  # deny mesh-to-dns policy is created only in egress flow
+                assert not route.is_internal_dest
+                deny_policy = self.init_deny_policy(vs, route_cnt, gtw)
```

Comment:
The deny policy is not implied by the virtual service that configures the connectivity from the gateway to the external destinations, but rather from the virtual service that is configured per `mesh` gateway and defines routing from mesh nodes to the egress gateway for the relevant hosts.. 
So this is not the place where the deny policy should be added? 
Here we did not yet check the existence of such virtual service? 
What if only one virtual service out of the 2 expected is defined (for the egress flow)?

Reply:
Why? I would rather say that deny policy is implied by the vs that configures the connectivity from the gateway to the external dns nodes (thus denying other peers to connect directly to those dns nodes).

Reply:
but the one actually configured for `mesh` is the one responsible for re-routing from internal nodes...
(also what if  only one virtual service out of the 2 expected (for the egress flow analysis) is defined?  )

Reply:
Added a check that the both parts of the egress flow are present